python_modules/libraries/dagster-dbt/dagster_dbt/asset_utils.py

```diff
@@ -680,255 +680,215 @@ def is_non_asset_node(dbt_resource_props: Mapping[str, Any]):
+        elif is_valid_upstream_node(parent_node_info):
+            upstreams.add(parent_unique_id)
+
+    return upstreams
+
+
+def get_asset_spec(
+    translator: "DagsterDbtTranslator",
+    manifest: Mapping[str, Any],
     dbt_nodes: Mapping[str, Any],
-    selected_unique_ids: AbstractSet[str],
-    asset_resource_types: List[str],
-) -> Mapping[str, FrozenSet[str]]:
-    def _valid_parent_node(dbt_resource_props):
-        # sources are valid parents, but not assets
-        return dbt_resource_props["resource_type"] in asset_resource_types + ["source"]
-
-    asset_deps: Dict[str, Set[str]] = {}
-    for unique_id in selected_unique_ids:
-        dbt_resource_props = dbt_nodes[unique_id]
-        node_resource_type = dbt_resource_props["resource_type"]
+    group_props: Mapping[str, Any],
+    project: Optional["DbtProject"],
+    resource_props: Mapping[str, Any],
+) -> AssetSpec:
+    """Returns an AssetSpec representing a specific dbt resource. In the future, this will be a method directly on
+    the DagsterDbtTranslator.
+    """
+    from dagster_dbt.dagster_dbt_translator import DbtManifestWrapper
 
-        # skip non-assets, such as semantic models, metrics, tests, and ephemeral models
-        if is_non_asset_node(dbt_resource_props) or node_resource_type not in asset_resource_types:
-            continue
+    # calculate the dependencies for the asset
+    upstream_ids = get_upstream_unique_ids(dbt_nodes, resource_props)
+    deps = [
+        AssetDep(
+            asset=translator.get_asset_key(dbt_nodes[upstream_id]),
+            partition_mapping=translator.get_partition_mapping(
+                resource_props, dbt_nodes[upstream_id]
+            ),
+        )
+        for upstream_id in upstream_ids
+    ]
+    self_partition_mapping = translator.get_partition_mapping(resource_props, resource_props)
+    if self_partition_mapping and has_self_dependency(resource_props):
+        deps.append(
+            AssetDep(
+                asset=translator.get_asset_key(resource_props),
+                partition_mapping=self_partition_mapping,
+            )
+        )
+    # base spec based off of default functions and user-defined functions
+    resource_group_props = group_props.get(resource_props.get("group") or "")
+    spec = AssetSpec(
+        key=translator.get_asset_key(resource_props),
+        deps=deps,
+        description=translator.get_description(resource_props),
+        metadata=translator.get_metadata(resource_props),
+        skippable=True,
+        group_name=translator.get_group_name(resource_props),
+        code_version=translator.get_code_version(resource_props),
+        automation_condition=translator.get_automation_condition(resource_props),
+        freshness_policy=translator.get_freshness_policy(resource_props),
+        owners=translator.get_owners(
+            {
+                **resource_props,
+                # this overrides the group key in resource_props, which is bad, but
```

Comment:
can you provide a bit more detail about exactly what breaks

Reply:
this is just me noting a pre-existing bad pattern, which is that we pass in an unstructured dictionary that is subtly different for this one specific method, which makes it hard for a user to write a custom function against (as if they're familiar with the dictionary structure this might surprise them)

updated comment